### ISSUE TITLE
Any palns for surpporting more conv kernel?

### ISSUE THREAD

**Issue by ZhangZelin-ustc** [pytorch/ao#483]
To speed up SD model more, will more conv kernel be supported?

**Comment by msaroufim** [comment]
Hi @ZhangZelin-ustc could you be more specific? Any specific convolution variants you're looking for us to accelerate?

**Comment by supriyar** [comment]
PyTorch core does have an int8 conv operator implementation (via cuDNN backend) here https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/quantized/cudnn/Conv.cpp. 

If there is interest, we could potentially hook this up to our `quantize_` API call in torchao to see how it performs. @ZhangZelin-ustc do you have a specific use case in mind that you want to speedup with conv?

cc @jerryzh168

**Comment by ZhangZelin-ustc** [comment]
@supriyar I would be very grateful for that.
Because of their widespread use, I think some [U-Nets](https://github.com/huggingface/diffusers/tree/main/src/diffusers/models/unets) in the 🤗 diffusers library are great cases.
For [SDXL](https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0), [unet_2d_condition](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/unets/unet_2d_condition.py) accounts for the vast majority of computations, while in [SVD-xt](https://huggingface.co/stabilityai/stable-video-diffusion-img2vid-xt), [UNetSpatioTemporalConditionModel](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/unets/unet_spatio_temporal_condition.py) plays this role.
A significant proportion of conv layers appear in [downsample](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/downsampling.py) and [upsample](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/upsampling.py) in both Unets.

**Comment by jerryzh168** [comment]
> @supriyar I would be very grateful for that. Because of their widespread use, I think some [U-Nets](https://github.com/huggingface/diffusers/tree/main/src/diffusers/models/unets?rgh-link-date=2024-07-10T05%3A46%3A57Z) in the 🤗 diffusers library are great cases. For [SDXL](https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0), [unet_2d_condition](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/unets/unet_2d_condition.py?rgh-link-date=2024-07-10T05%3A46%3A57Z) accounts for the vast majority of computations, while in [SVD-xt](https://huggingface.co/stabilityai/stable-video-diffusion-img2vid-xt), [UNetSpatioTemporalConditionModel](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/unets/unet_spatio_temporal_condition.py?rgh-link-date=2024-07-10T05%3A46%3A57Z) plays this role. A significant proportion of conv layers appear in [downsample](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/downsampling.py?rgh-link-date=2024-07-10T05%3A46%3A57Z) and [upsample](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/upsampling.py?rgh-link-date=2024-07-10T05%3A46%3A57Z) in both Unets.

implementing quantized conv kernel is not easy I think, is there any flavor of quantization that you are interested in, that already have proven accuracy results for your use cases? this is talking about: static/weight_only/dynamic, asymmetric/symmetric, bitwidth, per_tensor/per_channel for act/weight etc.

**Comment by ZhangZelin-ustc** [comment]
> > @supriyar I would be very grateful for that. Because of their widespread use, I think some [U-Nets](https://github.com/huggingface/diffusers/tree/main/src/diffusers/models/unets?rgh-link-date=2024-07-10T05%3A46%3A57Z) in the 🤗 diffusers library are great cases. For [SDXL](https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0), [unet_2d_condition](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/unets/unet_2d_condition.py?rgh-link-date=2024-07-10T05%3A46%3A57Z) accounts for the vast majority of computations, while in [SVD-xt](https://huggingface.co/stabilityai/stable-video-diffusion-img2vid-xt), [UNetSpatioTemporalConditionModel](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/unets/unet_spatio_temporal_condition.py?rgh-link-date=2024-07-10T05%3A46%3A57Z) plays this role. A significant proportion of conv layers appear in [downsample](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/downsampling.py?rgh-link-date=2024-07-10T05%3A46%3A57Z) and [upsample](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/upsampling.py?rgh-link-date=2024-07-10T05%3A46%3A57Z) in both Unets.
> 
> implementing quantized conv kernel is not easy I think, is there any flavor of quantization that you are interested in, that already have proven accuracy results for your use cases? this is talking about: static/weight_only/dynamic, asymmetric/symmetric, bitwidth, per_tensor/per_channel for act/weight etc.

So far, I have tried using [TRT-modelopt](https://github.com/NVIDIA/TensorRT-Model-Optimizer) to [quantify](https://github.com/NVIDIA/TensorRT-Model-Optimizer/tree/main/diffusers/quantization) SD and SVD. It provides a very high guarantee of accuracy. But it is partially closed-source, and all I can do is basically configure parameters on top of the interfaces it provides.

I used post-training static quantization (W8A8) to quantize the aforementioned models. Both weights and activations are 8-bit. Pre-channel quantization is applied to the weights, and per-tensor quantization is applied to the activations.It uses calibration to obtain the scale and amax, so I believe it is asymmetric quantization.

In fact, I almost completely configured according to the `get_int8_config` in [this file](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/diffusers/quantization/utils.py).

**Comment by ZhangZelin-ustc** [comment]
@gau-nernst Sure, on the NVIDIA L20, following its [README](https://github.com/NVIDIA/TensorRT-Model-Optimizer/tree/main/diffusers/quantization), the UNet of SDXL achieves an acceleration ratio of around 1.5, while the acceleration ratio of the UNet in SVD drops to around 1.1, and in our larger-scale models, there is almost no acceleration at all.

We suspect that some strategies of TRT introduce overhead that offsets the acceleration brought by quantization, because looking at the figure below, we can indeed confirm that the kernel-level acceleration ratio matches the theoretical value. However, due to the completely closed-source nature of TRT, there is not much additional we can do.

Note that these acceleration ratios are brought about by quantizing both linear and conv layers.

The figure below is a comparison of a certain conv layer kernel in a standard SDXL after int8 quantization.
![image](https://github.com/pytorch/ao/assets/49590336/bd828567-4d19-468a-9846-506950140d49)

**Comment by supriyar** [comment]
this might be relevant to the discussion here -  HF team has more details on quantizing diffusion model using int8 dynamic quantization using torchao (which involved replacing pointwise convs with linear layers) https://huggingface.co/docs/diffusers/en/tutorials/fast_diffusion#dynamic-quantization
cc @HDCharles

**Comment by ZhangZelin-ustc** [comment]
> this might be relevant to the discussion here -  HF team has more details on quantizing diffusion model using int8 dynamic quantization using torchao (which involved replacing pointwise convs with linear layers). The speedup is not as high as what you mention by quantizing conv layers though. https://huggingface.co/docs/diffusers/en/tutorials/fast_diffusion#dynamic-quantization
> cc @HDCharles 

In fact, I have tried using it before. It relies on an older version of torchao, but the changes needed are not too many.

An interesting phenomenon is that, contrary to TRT's acceleration ratio decreasing as the model size increases, its acceleration ratio increases with the model size. SD has negative acceleration, SDXL has no acceleration, and SVD has a certain acceleration ratio (seems to be 1.2, if I remember correctly).

**Comment by jeffhsu3** [comment]
Audio models and time-series models rely on Conv1Ds.  Though I think speed isn't as big of an issue, and more of consistency with edge devices:  https://huggingface.co/kyutai/mimi .

**Comment by jerryzh168** [comment]
we have some fp8 dynamic quant conv2d/3d supports for B200 now: https://github.com/pytorch/ao/blob/bedf37e55e4dd0d9554db05db8c8c016a819760c/test/quantization/quantize_/workflows/float8/test_float8_tensor.py#L406
